dev_guidelines: initiate a paragraph about unit-tests (#1156)
ansible-collections/amazon.aws: docs/docsite/rst/dev_guidelines.rst

Unsupported integration tests will not be automatically run by CI.  However, the
necessary policies should be available so that the tests can be manually run by
someone performing a PR review or writing a patch.

Unit-tests for AWS plugins
==========================

Why do we need unit-tests when we've got functional tests
---------------------------------------------------------

> Unit-tests are much faster and more suitable to test corner cases. They are also don't depend on a third party service

Reviewer: ```suggestion
Unit-tests are much faster and more suitable to test corner cases. They also don't depend on a third party service
```

Unsupported integration tests will not be automatically run by CI.  However, the
necessary policies should be available so that the tests can be manually run by
someone performing a PR review or writing a patch.

Unit-tests for AWS plugins
==========================

Why do we need unit-tests when we've got functional tests
---------------------------------------------------------

Unit-tests are much faster and more suitable to test corner cases. They also don't depend on a third party service

> and thus, a failure is less likley to be a false positive.

Reviewer 2: ```suggestion
and thus, a failure is less likely to be a false positive.
```

How to keep my code simple?
---------------------------

Ideally, you should break up your code in tiny functions. Each function should have a limited number of parameters
and a low amount of cross dependencies with the rest of the code (low coupling):

> - Don't pass a large data structure to a function if it only use one field. This clarify the inputs of your

Reviewer 3: ```suggestion
- Don't pass a large data structure to a function if it only use one field. This clarifies the inputs of your
```

> - Don't pass a large data structure to a function if it only use one field. This clarifies the inputs of your

Reviewer 2: ```suggestion
- Don't pass a large data structure to a function if it only uses one field. This clarifies the inputs of your
```

>   function (the contract) and also reduce the risk of an unexpected transformation of the data structure

Reviewer: ```suggestion
  function (the contract) and also reduces the risk of an unexpected transformation of the data structure
```

  from within the function.
- The boto client object is complex and can be source of unwanted side-effect. It's better to isolate the calls
  in dedicated functions. These functions will have their own unit-tests.
- Don't pass the ``module`` object when you only need the read a couple of parameters from ``module.params``.

>   Pass the parameter directly to your function. By doing so, you're explicite about what the function inputs

Reviewer 3: ```suggestion
  Pass the parameter directly to your function. By doing so, you're explicit about the function's inputs
```

  (the contract) and you reduce potential side-effect.

> Unit-tests guidlines

Reviewer: ```suggestion
Unit-tests guidelines
```

> --------------------

Reviewer: ```suggestion
---------------------
```

> Ideally, all the ``module_utils`` should be covered by unit-tests. However we acknoledge that writing unit-tests may

Reviewer 3: ```suggestion
Ideally, all the ``module_utils`` should be covered by unit-tests. However we acknowledge that writing unit-tests may
```

> be challenging and we also accept contribution with no coverage. The test coverage of the other type of plugins is recommand.

Reviewer 3: ```suggestion
be challenging and we also accept contribution with no coverage. Test coverage of the other type of plugins is recommended.
```

> 

Author: Hum, indeed. I've rephrased a bit the the sentence to clarify we only speak about unit-test.

> - Our tests are run with ``pytest`` and we use the features it provides such as Fixtures, Parametrization.

Reviewer 2: Wouldn't make sense to provide a generic example for Fixtures and Parametrization?

Author: Pytest documentation is pretty good and cover these features. But we can add some links pointing on was we identify as our best example of well written unit-tests.
This being said, I would suggest to do this in another patch.

Reviewer: The pytest docs are very good once you understand what you're looking at, but I found some of the examples a little too abstracted from reality to understand at first especially as it applies to our specific modules and module_utils.

> - The use of ``unittest.TestCase`` is discouraged for the sack of consistency and simplicity.

Reviewer 3: ```suggestion
- The use of ``unittest.TestCase`` is discouraged for the sake of consistency and simplicity.
```

Author: whoa, that was a nice one :-).

- Unit-tests should run fine without any network connection.
- It's not necessary to mock all the boto3/botocore calls (``get_paginator()``, ``paginate()``, etc). It's often better to just set-up a function that wrap these calls and mock the result.
- Simplicity prevails. Tests should be short and cover a limited set of features.

> Pytest is well documented and you will find some example in its `ow-to guides <https://docs.pytest.org/en/latest/how-to/index.html>`_

Reviewer 3: ```suggestion
Pytest is well documented and you will find some example in its `how-to guides <https://docs.pytest.org/en/latest/how-to/index.html>`_
```
